client/state/sites/hooks/use-premium-global-styles.ts

```diff
@@ -0,0 +1,103 @@
+import { isEnabled } from '@automattic/calypso-config';
+import { select } from '@wordpress/data';
+import { useQuery } from 'react-query';
+import { useSelector } from 'react-redux';
+import { ONBOARD_STORE } from 'calypso/landing/stepper/stores';
+import wpcom from 'calypso/lib/wp';
+import { getSite } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+type GlobalStylesStatus = {
+	shouldLimitGlobalStyles: boolean;
+	globalStylesInUse: boolean;
+};
+
+type Site = {
+	isCreated: boolean;
+	id: number;
+};
+
+type Onboard = {
+	stepProgress?: {
+		count: number;
+		progress: number;
+	};
+};
+
+const NEW_SITE_GLOBAL_STYLES_INFO: GlobalStylesStatus = {
+	// The next line should be replaced with true once the Gating global styles feature is live.
+	// That will make all non-created sites to limit global styles.
+	shouldLimitGlobalStyles: isEnabled( 'limit-global-styles' ),
+	globalStylesInUse: false,
+};
+
+// While we are loading the Global Styles Info we can't assume that we should limit global styles, or we would be
+// showing notices for paid sites until we fetch the data from the server.
+const DEFAULT_GLOBAL_STYLES_INFO: GlobalStylesStatus = {
+	shouldLimitGlobalStyles: false,
+	globalStylesInUse: false,
+};
+
+const getGSInfoEndpointForSite = ( siteId: number ) => ( {
+	path: `sites/${ siteId }/global-styles/status`,
+	apiNamespace: 'wpcom/v2/',
+} );
+
+const mapSiteIdToSite = ( siteId: number | null ) => ( {
+	isCreated: !! siteId,
+	id: siteId ?? 0,
+} );
+
+const mapResponseToGlobalStylesStatus = ( data: GlobalStylesStatus ): GlobalStylesStatus => ( {
+	...DEFAULT_GLOBAL_STYLES_INFO,
+	...data,
+} );
+
+const isOnboarding = ( onboard: Onboard ): boolean => onboard?.stepProgress !== undefined;
+
+const getGlobalStylesInfoForSite = ( site: Site ): GlobalStylesStatus => {
+	if ( ! site.isCreated ) {
+		return NEW_SITE_GLOBAL_STYLES_INFO;
+	}
+
+	const globalStylesInfoEndpoint = getGSInfoEndpointForSite( site.id );
+	return wpcom.req
+		.get( globalStylesInfoEndpoint )
+		.then( mapResponseToGlobalStylesStatus )
+		.catch( () => DEFAULT_GLOBAL_STYLES_INFO );
```

Comment:
The benefit of having semantics instead of actual code is that you don't need to read how something is implemented to know what it's doing. In this specific case, you can see that we are mapping the API response to GlobalStylesStatus without reading how it's implemented (you can choose to read the implementation but aren't forced to). 

With good implementations of indirection, you don't need to remember three functions, you also don't need to navigate to understand the behavior, you just need to read 3 lines of code instead of 60. The words used should be enough to explain what is going on. Indirection could be a problem when it adds no semantic value to the code.

As an example look at the way we're determining the site's information, you need to read 15 lines of code, and understand them before you can establish that we're determining if there is a siteId, and if not, the site hasn't been created. If we have just used a function here we would just need to read one line of code, and the responsibilities of the main hook would also be reduced in scope.

I doubt we will ever reach an agreement since this is completely subjective, that's why I usually don't comment on things like this in your code reviews, I know I'm not going to be able to show you/convince you of the value I see in it, the same way you won't be able to convince me the value of de-structured code.

Reply:
> I know I'm not going to be able to show you/convince you of the value I see in it, the same way you won't be able to convince me the value of de-structured code.

I was convinced at some point! At the beginning I preferred to abstract/hide the implementation details, but then someone gave me feedback about not hiding them and I re-considered my guidelines. Which is why I gave this feedback to you now. But it's ok if you think you'll never be convinced. I'll stop providing this kind of feedback from now on.

Reply:
I see, even though I'll never be convinced about that particular way of coding being correct or superior in any way (promotes unmaintainable, non-reusable, untestable code, and increases the cognitive load for understanding it), I can adapt to it.

I've changed the code.

Reply:
Already discussed privately on Slack, but also wanted to leave a comment here to close the loop. It was not my intention to impose a code style, but to bring up a different perspective for you to consider. If you don't see value in that perspective, it's fine to keep the implementation with the previous approach. This is something subjective as you pointed out, so you're not expected to disagree and commit.